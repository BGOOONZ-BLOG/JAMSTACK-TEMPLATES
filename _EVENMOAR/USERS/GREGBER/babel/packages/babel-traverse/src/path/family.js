// This file contains methods responsible for dealing with/retrieving children or siblings.

import type TraversalContext from "../index";
import NodePath from "./index";
import * as t from "@babel/types";

export function getOpposite(): ?NodePath {
  if (this.key === "left") {
    return this.getSibling("right");
  } else if (this.key === "right") {
    return this.getSibling("left");
  }
}

function addCompletionRecords(path, paths) {
  if (path) return paths.concat(path.getCompletionRecords());
  return paths;
}

export function getCompletionRecords(): Array {
  let paths = [];

  if (this.isIfStatement()) {
    paths = addCompletionRecords(this.get("consequent"), paths);
    paths = addCompletionRecords(this.get("alternate"), paths);
  } else if (this.isDoExpression() || this.isFor() || this.isWhile()) {
    paths = addCompletionRecords(this.get("body"), paths);
  } else if (this.isProgram() || this.isBlockStatement()) {
    paths = addCompletionRecords(this.get("body").pop(), paths);
  } else if (this.isFunction()) {
    return this.get("body").getCompletionRecords();
  } else if (this.isTryStatement()) {
    paths = addCompletionRecords(this.get("block"), paths);
    paths = addCompletionRecords(this.get("handler"), paths);
    paths = addCompletionRecords(this.get("finalizer"), paths);
  } else if (this.isCatchClause()) {
    paths = addCompletionRecords(this.get("body"), paths);
  } else {
    paths.push(this);
  }

  return paths;
}

export function getSibling(key): NodePath {
  return NodePath.get({
    parentPath: this.parentPath,
    parent: this.parent,
    container: this.container,
    listKey: this.listKey,
    key: key,
  });
}

export function getPrevSibling(): NodePath {
  return this.getSibling(this.key - 1);
}

export function getNextSibling(): NodePath {
  return this.getSibling(this.key + 1);
}

export function getAllNextSiblings(): Array<NodePath> {
  let _key = this.key;
  let sibling: NodePath = this.getSibling(++_key);
  const siblings: Array<NodePath> = [];
  while (sibling.node) {
    siblings.push(sibling);
    sibling = this.getSibling(++_key);
  }
  return siblings;
}

export function getAllPrevSiblings(): Array<NodePath> {
  let _key = this.key;
  let sibling: NodePath = this.getSibling(--_key);
  const siblings: Array<NodePath> = [];
  while (sibling.node) {
    siblings.push(sibling);
    sibling = this.getSibling(--_key);
  }
  return siblings;
}

export function get(
  key: string,
  context?: boolean | TraversalContext,
): NodePath {
  if (context === true) context = this.context;
  const parts = key.split(".");
  if (parts.length === 1) {
    // "foo"
    return this._getKey(key, context);
  } else {
    // "foo.bar"
    return this._getPattern(parts, context);
  }
}

export function _getKey(key, context?) {
  const node = this.node;
  const container = node[key];

  if (Array.isArray(container)) {
    // requested a container so give them all the paths
    return container.map((_, i) => {
      return NodePath.get({
        listKey: key,
        parentPath: this,
        parent: node,
        container: container,
        key: i,
      }).setContext(context);
    });
  } else {
    return NodePath.get({
      parentPath: this,
      parent: node,
      container: node,
      key: key,
    }).setContext(context);
  }
}

export function _getPattern(parts, context) {
  let path = this;
  for (const part of (parts: Array)) {
    if (part === ".") {
      path = path.parentPath;
    } else {
      if (Array.isArray(path)) {
        path = path[part];
      } else {
        path = path.get(part, context);
      }
    }
  }
  return path;
}

export function getBindingIdentifiers(duplicates?): Object {
  return t.getBindingIdentifiers(this.node, duplicates);
}

export function getOuterBindingIdentifiers(duplicates?): Object {
  return t.getOuterBindingIdentifiers(this.node, duplicates);
}

// original source - https://github.com/babel/babel/blob/master/packages/babel-types/src/retrievers.js
// path.getBindingIdentifiers returns nodes where the following re-implementation
// returns paths
export function getBindingIdentifierPaths(
  duplicates = false,
  outerOnly = false,
) {
  const path = this;
  let search = [].concat(path);
  const ids = Object.create(null);

  while (search.length) {
    const id = search.shift();
    if (!id) continue;
    if (!id.node) continue;

    const keys = t.getBindingIdentifiers.keys[id.node.type];

    if (id.isIdentifier()) {
      if (duplicates) {
        const _ids = (ids[id.node.name] = ids[id.node.name] || []);
        _ids.push(id);
      } else {
        ids[id.node.name] = id;
      }
      continue;
    }

    if (id.isExportDeclaration()) {
      const declaration = id.get("declaration");
      if (declaration.isDeclaration()) {
        search.push(declaration);
      }
      continue;
    }

    if (outerOnly) {
      if (id.isFunctionDeclaration()) {
        search.push(id.get("id"));
        continue;
      }
      if (id.isFunctionExpression()) {
        continue;
      }
    }

    if (keys) {
      for (let i = 0; i < keys.length; i++) {
        const key = keys[i];
        const child = id.get(key);
        if (Array.isArray(child) || child.node) {
          search = search.concat(child);
        }
      }
    }
  }

  return ids;
}

export function getOuterBindingIdentifierPaths(duplicates?) {
  return this.getBindingIdentifierPaths(duplicates, true);
}
