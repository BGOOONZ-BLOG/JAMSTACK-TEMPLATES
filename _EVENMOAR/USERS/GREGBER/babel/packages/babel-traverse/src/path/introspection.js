// This file contains methods responsible for introspecting the current path for certain values.

import type NodePath from "./index";
import includes from "lodash/includes";
import * as t from "@babel/types";

/**
 * Match the current node if it matches the provided `pattern`.
 *
 * For example, given the match `React.createClass` it would match the
 * parsed nodes of `React.createClass` and `React["createClass"]`.
 */

export function matchesPattern(
  pattern: string,
  allowPartial?: boolean,
): boolean {
  return t.matchesPattern(this.node, pattern, allowPartial);
}

/**
 * Check whether we have the input `key`. If the `key` references an array then we check
 * if the array has any items, otherwise we just check if it's falsy.
 */

export function has(key): boolean {
  const val = this.node && this.node[key];
  if (val && Array.isArray(val)) {
    return !!val.length;
  } else {
    return !!val;
  }
}

/**
 * Description
 */

export function isStatic() {
  return this.scope.isStatic(this.node);
}

/**
 * Alias of `has`.
 */

export const is = has;

/**
 * Opposite of `has`.
 */

export function isnt(key): boolean {
  return !this.has(key);
}

/**
 * Check whether the path node `key` strict equals `value`.
 */

export function equals(key, value): boolean {
  return this.node[key] === value;
}

/**
 * Check the type against our stored internal type of the node. This is handy when a node has
 * been removed yet we still internally know the type and need it to calculate node replacement.
 */

export function isNodeType(type: string): boolean {
  return t.isType(this.type, type);
}

/**
 * This checks whether or not we're in one of the following positions:
 *
 *   for (KEY in right);
 *   for (KEY;;);
 *
 * This is because these spots allow VariableDeclarations AND normal expressions so we need
 * to tell the path replacement that it's ok to replace this with an expression.
 */

export function canHaveVariableDeclarationOrExpression() {
  return (
    (this.key === "init" || this.key === "left") && this.parentPath.isFor()
  );
}

/**
 * This checks whether we are swapping an arrow function's body between an
 * expression and a block statement (or vice versa).
 *
 * This is because arrow functions may implicitly return an expression, which
 * is the same as containing a block statement.
 */

export function canSwapBetweenExpressionAndStatement(replacement) {
  if (this.key !== "body" || !this.parentPath.isArrowFunctionExpression()) {
    return false;
  }

  if (this.isExpression()) {
    return t.isBlockStatement(replacement);
  } else if (this.isBlockStatement()) {
    return t.isExpression(replacement);
  }

  return false;
}

/**
 * Check whether the current path references a completion record
 */

export function isCompletionRecord(allowInsideFunction?) {
  let path = this;
  let first = true;

  do {
    const container = path.container;

    // we're in a function so can't be a completion record
    if (path.isFunction() && !first) {
      return !!allowInsideFunction;
    }

    first = false;

    // check to see if we're the last item in the container and if we are
    // we're a completion record!
    if (Array.isArray(container) && path.key !== container.length - 1) {
      return false;
    }
  } while ((path = path.parentPath) && !path.isProgram());

  return true;
}

/**
 * Check whether or not the current `key` allows either a single statement or block statement
 * so we can explode it if necessary.
 */

export function isStatementOrBlock() {
  if (
    this.parentPath.isLabeledStatement() ||
    t.isBlockStatement(this.container)
  ) {
    return false;
  } else {
    return includes(t.STATEMENT_OR_BLOCK_KEYS, this.key);
  }
}

/**
 * Check if the currently assigned path references the `importName` of `moduleSource`.
 */

export function referencesImport(moduleSource, importName) {
  if (!this.isReferencedIdentifier()) return false;

  const binding = this.scope.getBinding(this.node.name);
  if (!binding || binding.kind !== "module") return false;

  const path = binding.path;
  const parent = path.parentPath;
  if (!parent.isImportDeclaration()) return false;

  // check moduleSource
  if (parent.node.source.value === moduleSource) {
    if (!importName) return true;
  } else {
    return false;
  }

  if (path.isImportDefaultSpecifier() && importName === "default") {
    return true;
  }

  if (path.isImportNamespaceSpecifier() && importName === "*") {
    return true;
  }

  if (path.isImportSpecifier() && path.node.imported.name === importName) {
    return true;
  }

  return false;
}

/**
 * Get the source code associated with this node.
 */

export function getSource() {
  const node = this.node;
  if (node.end) {
    const code = this.hub.getCode();
    if (code) return code.slice(node.start, node.end);
  }
  return "";
}

export function willIMaybeExecuteBefore(target) {
  return this._guessExecutionStatusRelativeTo(target) !== "after";
}

/**
 * Given a `target` check the execution status of it relative to the current path.
 *
 * "Execution status" simply refers to where or not we **think** this will execuete
 * before or after the input `target` element.
 */

export function _guessExecutionStatusRelativeTo(target) {
  // check if the two paths are in different functions, we can't track execution of these
  const targetFuncParent =
    target.scope.getFunctionParent() || target.scope.getProgramParent();
  const selfFuncParent =
    this.scope.getFunctionParent() || target.scope.getProgramParent();

  // here we check the `node` equality as sometimes we may have different paths for the
  // same node due to path thrashing
  if (targetFuncParent.node !== selfFuncParent.node) {
    const status = this._guessExecutionStatusRelativeToDifferentFunctions(
      targetFuncParent,
    );
    if (status) {
      return status;
    } else {
      target = targetFuncParent.path;
    }
  }

  const targetPaths = target.getAncestry();
  if (targetPaths.indexOf(this) >= 0) return "after";

  const selfPaths = this.getAncestry();

  // get ancestor where the branches intersect
  let commonPath;
  let targetIndex;
  let selfIndex;
  for (selfIndex = 0; selfIndex < selfPaths.length; selfIndex++) {
    const selfPath = selfPaths[selfIndex];
    targetIndex = targetPaths.indexOf(selfPath);
    if (targetIndex >= 0) {
      commonPath = selfPath;
      break;
    }
  }
  if (!commonPath) {
    return "before";
  }

  // get the relationship paths that associate these nodes to their common ancestor
  const targetRelationship = targetPaths[targetIndex - 1];
  const selfRelationship = selfPaths[selfIndex - 1];
  if (!targetRelationship || !selfRelationship) {
    return "before";
  }

  // container list so let's see which one is after the other
  if (
    targetRelationship.listKey &&
    targetRelationship.container === selfRelationship.container
  ) {
    return targetRelationship.key > selfRelationship.key ? "before" : "after";
  }

  // otherwise we're associated by a parent node, check which key comes before the other
  const keys = t.VISITOR_KEYS[commonPath.type];
  const targetKeyPosition = keys.indexOf(targetRelationship.key);
  const selfKeyPosition = keys.indexOf(selfRelationship.key);
  return targetKeyPosition > selfKeyPosition ? "before" : "after";
}

export function _guessExecutionStatusRelativeToDifferentFunctions(
  targetFuncParent,
) {
  const targetFuncPath = targetFuncParent.path;
  if (!targetFuncPath.isFunctionDeclaration()) return;

  // so we're in a completely different function, if this is a function declaration
  // then we can be a bit smarter and handle cases where the function is either
  // a. not called at all (part of an export)
  // b. called directly
  const binding = targetFuncPath.scope.getBinding(targetFuncPath.node.id.name);

  // no references!
  if (!binding.references) return "before";

  const referencePaths: Array<NodePath> = binding.referencePaths;

  // verify that all of the references are calls
  for (const path of referencePaths) {
    if (path.key !== "callee" || !path.parentPath.isCallExpression()) {
      return;
    }
  }

  let allStatus;

  // verify that all the calls have the same execution status
  for (const path of referencePaths) {
    // if a reference is a child of the function we're checking against then we can
    // safelty ignore it
    const childOfFunction = !!path.find(
      path => path.node === targetFuncPath.node,
    );
    if (childOfFunction) continue;

    const status = this._guessExecutionStatusRelativeTo(path);

    if (allStatus) {
      if (allStatus !== status) return;
    } else {
      allStatus = status;
    }
  }

  return allStatus;
}

/**
 * Resolve a "pointer" `NodePath` to it's absolute path.
 */

export function resolve(dangerous, resolved) {
  return this._resolve(dangerous, resolved) || this;
}

export function _resolve(dangerous?, resolved?): ?NodePath {
  // detect infinite recursion
  // todo: possibly have a max length on this just to be safe
  if (resolved && resolved.indexOf(this) >= 0) return;

  // we store all the paths we've "resolved" in this array to prevent infinite recursion
  resolved = resolved || [];
  resolved.push(this);

  if (this.isVariableDeclarator()) {
    if (this.get("id").isIdentifier()) {
      return this.get("init").resolve(dangerous, resolved);
    } else {
      // otherwise it's a request for a pattern and that's a bit more tricky
    }
  } else if (this.isReferencedIdentifier()) {
    const binding = this.scope.getBinding(this.node.name);
    if (!binding) return;

    // reassigned so we can't really resolve it
    if (!binding.constant) return;

    // todo - lookup module in dependency graph
    if (binding.kind === "module") return;

    if (binding.path !== this) {
      const ret = binding.path.resolve(dangerous, resolved);
      // If the identifier resolves to parent node then we can't really resolve it.
      if (this.find(parent => parent.node === ret.node)) return;
      return ret;
    }
  } else if (this.isTypeCastExpression()) {
    return this.get("expression").resolve(dangerous, resolved);
  } else if (dangerous && this.isMemberExpression()) {
    // this is dangerous, as non-direct target assignments will mutate it's state
    // making this resolution inaccurate

    const targetKey = this.toComputedKey();
    if (!t.isLiteral(targetKey)) return;

    const targetName = targetKey.value;

    const target = this.get("object").resolve(dangerous, resolved);

    if (target.isObjectExpression()) {
      const props = target.get("properties");
      for (const prop of (props: Array)) {
        if (!prop.isProperty()) continue;

        const key = prop.get("key");

        // { foo: obj }
        let match =
          prop.isnt("computed") && key.isIdentifier({ name: targetName });

        // { "foo": "obj" } or { ["foo"]: "obj" }
        match = match || key.isLiteral({ value: targetName });

        if (match) return prop.get("value").resolve(dangerous, resolved);
      }
    } else if (target.isArrayExpression() && !isNaN(+targetName)) {
      const elems = target.get("elements");
      const elem = elems[targetName];
      if (elem) return elem.resolve(dangerous, resolved);
    }
  }
}

export function isConstantExpression() {
  if (this.isIdentifier()) {
    const binding = this.scope.getBinding(this.node.name);
    if (!binding) return false;
    return binding.constant;
  }

  if (this.isLiteral()) {
    if (this.isRegExpLiteral()) {
      return false;
    }

    if (this.isTemplateLiteral()) {
      return this.get("expressions").every(expression =>
        expression.isConstantExpression(),
      );
    }

    return true;
  }

  if (this.isUnaryExpression()) {
    if (this.get("operator").node !== "void") {
      return false;
    }

    return this.get("argument").isConstantExpression();
  }

  if (this.isBinaryExpression()) {
    return (
      this.get("left").isConstantExpression() &&
      this.get("right").isConstantExpression()
    );
  }

  return false;
}

export function isInStrictMode() {
  const start = this.isProgram() ? this : this.parentPath;

  const strictParent = start.find(path => {
    if (path.isProgram({ sourceType: "module" })) return true;

    if (path.isClass()) return true;

    if (!path.isProgram() && !path.isFunction()) return false;

    if (
      path.isArrowFunctionExpression() &&
      !path.get("body").isBlockStatement()
    ) {
      return false;
    }

    let { node } = path;
    if (path.isFunction()) node = node.body;

    for (const directive of node.directives) {
      if (directive.value.value === "use strict") {
        return true;
      }
    }
  });

  return !!strictParent;
}
