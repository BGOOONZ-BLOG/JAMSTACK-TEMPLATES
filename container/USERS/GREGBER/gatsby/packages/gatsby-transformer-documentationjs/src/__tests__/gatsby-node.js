import groupBy from "lodash/groupBy"
import path from "path"
import gatsbyNode from "../gatsby-node"

describe(`gatsby-transformer-documentationjs: onCreateNode`, () => {
  let createdNodes, updatedNodes
  const createNodeId = jest.fn(id => id)
  const createContentDigest = jest.fn().mockReturnValue(`content-digest`)

  const getFileNode = fixture => {
    return {
      id: `node_1`,
      children: [],
      absolutePath: path.join(__dirname, `fixtures`, fixture),
      internal: {
        mediaType: `application/javascript`,
        type: `File`,
      },
    }
  }

  const actions = {
    createNode: jest.fn(n => createdNodes.push(n)),
    createParentChildLink: jest.fn(n => {
      updatedNodes.push(n)
      const parentNode = createdNodes.find(node => node.id === n.parent.id)
      if (parentNode) {
        parentNode.children.push(n.child.id)
      } else if (n.parent.id !== `node_1`) {
        throw new Error(`Creating parent-child link for not existing parent`)
      }
    }),
  }

  const run = async (node = node, opts = {}) => {
    createdNodes = []
    updatedNodes = []
    await gatsbyNode.onCreateNode(
      {
        node,
        actions,
        createNodeId,
        createContentDigest,
      },
      opts
    )
  }

  describe(`Simple example`, () => {
    beforeAll(async () => {
      await run(getFileNode(`code.js`))
    })

    it(`creates doc json apple node`, () => {
      const appleNode = createdNodes.find(node => node.name === `apple`)
      expect(appleNode).toBeDefined()
    })

    it(`should extract out a description, params, and examples`, () => {
      const appleNode = createdNodes.find(node => node.name === `apple`)

      expect(appleNode.examples.length).toBe(1)
      expect(appleNode.examples[0]).toMatchSnapshot(`example`)

      const appleDescriptionNode = createdNodes.find(
        node => node.id === appleNode.description___NODE
      )

      expect(appleDescriptionNode).toBeDefined()
      expect(appleDescriptionNode.internal.content).toMatchSnapshot(
        `description content`
      )

      const paramNode = createdNodes.find(
        node => node.id === appleNode.params___NODE[0]
      )

      expect(paramNode).toBeDefined()
      expect(paramNode.name).toMatchSnapshot(`param name`)

      const paramDescriptionNode = createdNodes.find(
        node => node.id === paramNode.description___NODE
      )

      expect(paramDescriptionNode).toBeDefined()
      expect(paramDescriptionNode.internal.content).toMatchSnapshot(
        `param description`
      )
    })

    it(`should extract code and docs location`, () => {
      const appleNode = createdNodes.find(node => node.name === `apple`)

      expect(appleNode.docsLocation).toBeDefined()
      expect(appleNode.docsLocation).toEqual(
        expect.objectContaining({
          start: expect.objectContaining({
            line: 1,
          }),
          end: expect.objectContaining({
            line: 7,
          }),
        })
      )

      expect(appleNode.codeLocation).toBeDefined()
      expect(appleNode.codeLocation).toEqual(
        expect.objectContaining({
          start: expect.objectContaining({
            line: 8,
          }),
          end: expect.objectContaining({
            line: 10,
          }),
        })
      )
    })

    it(`doesn't create multiple nodes with same id`, () => {
      const groupedById = groupBy(createdNodes, `id`)
      Object.keys(groupedById).forEach(id =>
        expect(groupedById[id].length).toBe(1)
      )
    })
  })

  describe(`Complex example`, () => {
    beforeAll(async () => {
      await run(getFileNode(`complex-example.js`))
    })

    let callbackNode, typedefNode

    it(`should create top-level node for callback`, () => {
      callbackNode = createdNodes.find(
        node =>
          node.name === `CallbackType` &&
          node.kind === `typedef` &&
          node.parent === `node_1`
      )
      expect(callbackNode).toBeDefined()
    })

    describe(`should handle typedefs`, () => {
      it(`should create top-level node for typedef`, () => {
        typedefNode = createdNodes.find(
          node =>
            node.name === `ObjectType` &&
            node.kind === `typedef` &&
            node.parent === `node_1`
        )
        expect(typedefNode).toBeDefined()
      })

      let readyNode, nestedNode

      it(`should have property nodes for typedef`, () => {
        expect(typedefNode.properties___NODE).toBeDefined()
        expect(typedefNode.properties___NODE.length).toBe(2)
        ;[readyNode, nestedNode] = typedefNode.properties___NODE.map(paramID =>
          createdNodes.find(node => node.id === paramID)
        )
      })

      it(`should handle type applications`, () => {
        expect(readyNode).toMatchSnapshot()
      })

      let nestedFooNode, nestedOptionalNode, nestedCallbackNode

      it(`should have second param as nested object`, () => {
        expect(nestedNode.name).toBe(`nested`)
        expect(nestedNode.properties___NODE).toBeDefined()
        expect(nestedNode.properties___NODE.length).toBe(3)
        ;[
          nestedFooNode,
          nestedOptionalNode,
          nestedCallbackNode,
        ] = nestedNode.properties___NODE.map(paramID =>
          createdNodes.find(node => node.id === paramID)
        )
      })

      it(`should strip prefixes from nested nodes`, () => {
        expect(nestedFooNode.name).not.toContain(`nested`)
        expect(nestedFooNode.name).toEqual(`foo`)
      })

      it(`should handle optional types`, () => {
        expect(nestedOptionalNode.name).toEqual(`optional`)
        expect(nestedOptionalNode.optional).toEqual(true)
        expect(nestedOptionalNode.type).toEqual(
          expect.objectContaining({
            name: `number`,
            type: `NameExpression`,
          })
        )
      })

      it(`should handle typedefs in nested properties`, () => {
        expect(nestedCallbackNode.name).toEqual(`callback`)
        expect(nestedCallbackNode.optional).toEqual(false)
        expect(nestedCallbackNode.type).toEqual(
          expect.objectContaining({
            name: `CallbackType`,
            type: `NameExpression`,
            typeDef___NODE: callbackNode.id,
          })
        )
      })
    })

    describe(`should handle members`, () => {
      let complexNode, memberNode
      beforeAll(() => {
        complexNode = createdNodes.find(
          node => node.name === `complex` && node.parent === `node_1`
        )
      })

      it(`should create top-level node for complex type`, () => {
        expect(complexNode).toBeDefined()
      })

      it(`should have link from complex node to its members`, () => {
        expect(complexNode.members).toBeDefined()
        expect(complexNode.members.static___NODE).toBeDefined()
        expect(complexNode.members.static___NODE.length).toBe(1)

        memberNode = createdNodes.find(
          node => node.id === complexNode.members.static___NODE[0]
        )
        expect(memberNode).toBeDefined()
        expect(memberNode.parent).toEqual(complexNode.id)
      })

      it(`should handle type unions`, () => {
        expect(memberNode.type).toMatchSnapshot()
      })

      it(`should link to type to type definition`, () => {
        const typeElement = memberNode.type.elements.find(
          type => type.name === `ObjectType`
        )
        expect(typeElement.typeDef___NODE).toBe(typedefNode.id)
      })
    })

    it(`doesn't create multiple nodes with same id`, () => {
      const groupedById = groupBy(createdNodes, `id`)
      Object.keys(groupedById).forEach(id =>
        expect(groupedById[id].length).toBe(1)
      )
    })
  })

  describe(`Free floating comments`, () => {
    beforeAll(async () => {
      await run(getFileNode(`free-floating.js`))
    })

    it(`doesn't create multiple nodes with same id`, () => {
      const groupedById = groupBy(createdNodes, `id`)
      Object.keys(groupedById).forEach(id =>
        expect(groupedById[id].length).toBe(1)
      )
    })
  })

  describe(`JSDoc + Flow`, () => {
    beforeAll(async () => {
      await run(getFileNode(`jsdoc-and-flow.js`))
    })

    it(`merges JsDoc and flow types`, () => {
      const firstParamNodes = createdNodes.filter(node => node.name === `$0`)

      expect(firstParamNodes.length).toBe(1)
      expect(firstParamNodes[0].properties___NODE.length).toBe(1)

      const subField = createdNodes.find(
        node => node.id === firstParamNodes[0].properties___NODE[0]
      )

      expect(subField).toBeDefined()
      expect(subField.name).toBe(`nodeId`)
      expect(subField.type).toEqual(
        expect.objectContaining({
          name: `string`,
          type: `NameExpression`,
        })
      )
    })

    it(`flow: required param result in same data shape as JsDoc`, () => {
      const requiredFlowParam = createdNodes.find(
        node => node.name === `requiredParam`
      )

      expect(requiredFlowParam).toBeDefined()
      expect(requiredFlowParam).toEqual(
        expect.objectContaining({
          name: `requiredParam`,
          optional: false,
          type: expect.objectContaining({
            name: `string`,
            type: `NameExpression`,
          }),
        })
      )
    })

    it(`flow: optional param result in same data shape as JsDoc`, () => {
      const optionalFlowParam = createdNodes.find(
        node => node.name === `optionalParam`
      )

      expect(optionalFlowParam).toBeDefined()
      expect(optionalFlowParam).toEqual(
        expect.objectContaining({
          name: `optionalParam`,
          optional: true,
          type: expect.objectContaining({
            name: `number`,
            type: `NameExpression`,
          }),
        })
      )
    })
  })

  describe(`Sanity checks`, () => {
    it(`should extract create description nodes with markdown types`, () => {
      let types = groupBy(createdNodes, `internal.type`)
      expect(
        types.DocumentationJSComponentDescription.every(
          d => d.internal.mediaType === `text/markdown`
        )
      ).toBe(true)
    })

    it(`creates parent nodes before children`, () => {
      const seenNodes = []
      createdNodes.forEach(node => {
        seenNodes.push(node.id)

        node.children.forEach(childID => {
          expect(seenNodes.includes(childID)).not.toBe(true)
        })
      })
    })

    it(`should only process javascript File nodes`, async () => {
      await run({ internal: { mediaType: `text/x-foo` } })
      expect(createdNodes.length).toBe(0)

      await run({ internal: { mediaType: `application/javascript` } })
      expect(createdNodes.length).toBe(0)

      await run(getFileNode(`code.js`))
      expect(createdNodes.length).toBeGreaterThan(0)
    })
  })

  it(`doesn't cause a stack overflow for nodes of the same name`, () => {
    expect(run(getFileNode(`same-name.ts`))).resolves.toBeUndefined()
  })
})
