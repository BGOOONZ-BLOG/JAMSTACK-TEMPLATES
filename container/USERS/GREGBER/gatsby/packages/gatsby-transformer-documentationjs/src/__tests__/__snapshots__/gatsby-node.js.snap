// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatsby-transformer-documentationjs: onCreateNode Complex example should handle members should handle type unions 1`] = `
Object {
  "elements": Array [
    Object {
      "name": "ObjectType",
      "type": "NameExpression",
      "typeDef___NODE": "documentationJS node_1 path #[{\\"name\\":\\"ObjectType\\",\\"kind\\":\\"typedef\\"}] line 12",
    },
    Object {
      "name": "Object",
      "type": "NameExpression",
      "typeDef___NODE": null,
    },
  ],
  "type": "UnionType",
}
`;

exports[`gatsby-transformer-documentationjs: onCreateNode Complex example should handle typedefs should handle type applications 1`] = `
Object {
  "children": Array [
    "documentationJS documentationJS node_1 path #[{\\"name\\":\\"ObjectType\\",\\"kind\\":\\"typedef\\"}] line 12 path #[{\\"name\\":\\"ObjectType\\",\\"kind\\":\\"typedef\\"},{\\"fieldName\\":\\"properties\\",\\"fieldIndex\\":0}] line 3--DocumentationJSComponentDescription--comment.description",
  ],
  "commentNumber": null,
  "description___NODE": "documentationJS documentationJS node_1 path #[{\\"name\\":\\"ObjectType\\",\\"kind\\":\\"typedef\\"}] line 12 path #[{\\"name\\":\\"ObjectType\\",\\"kind\\":\\"typedef\\"},{\\"fieldName\\":\\"properties\\",\\"fieldIndex\\":0}] line 3--DocumentationJSComponentDescription--comment.description",
  "id": "documentationJS documentationJS node_1 path #[{\\"name\\":\\"ObjectType\\",\\"kind\\":\\"typedef\\"}] line 12 path #[{\\"name\\":\\"ObjectType\\",\\"kind\\":\\"typedef\\"},{\\"fieldName\\":\\"properties\\",\\"fieldIndex\\":0}] line 3",
  "internal": Object {
    "contentDigest": "content-digest",
    "type": "DocumentationJs",
  },
  "level": 1,
  "name": "ready",
  "optional": false,
  "parent": "documentationJS node_1 path #[{\\"name\\":\\"ObjectType\\",\\"kind\\":\\"typedef\\"}] line 12",
  "type": Object {
    "applications": Array [
      Object {
        "name": "any",
        "type": "NameExpression",
        "typeDef___NODE": null,
      },
    ],
    "expression": Object {
      "name": "Promise",
      "type": "NameExpression",
      "typeDef___NODE": null,
    },
    "type": "TypeApplication",
  },
}
`;

exports[`gatsby-transformer-documentationjs: onCreateNode Simple example should extract out a description, params, and examples: description content 1`] = `
"A pretty cool jsdoc example
"
`;

exports[`gatsby-transformer-documentationjs: onCreateNode Simple example should extract out a description, params, and examples: example 1`] = `
Object {
  "description": "const apple = require('apple')
apple()",
  "highlighted": "<span class=\\"token keyword\\">const</span> apple <span class=\\"token operator\\">=</span> <span class=\\"token function\\">require</span><span class=\\"token punctuation\\">(</span><span class=\\"token string\\">'apple'</span><span class=\\"token punctuation\\">)</span>
<span class=\\"token function\\">apple</span><span class=\\"token punctuation\\">(</span><span class=\\"token punctuation\\">)</span>",
  "raw": "const apple = require('apple')
apple()",
}
`;

exports[`gatsby-transformer-documentationjs: onCreateNode Simple example should extract out a description, params, and examples: param description 1`] = `
"A nice crispy apple
"
`;

exports[`gatsby-transformer-documentationjs: onCreateNode Simple example should extract out a description, params, and examples: param name 1`] = `"paramName"`;
